### DISCUSSION TITLE
Logistics of splitting the repository

### DISCUSSION THREAD

**Discussion by rossabaker** [http4s/http4s#4140] [Ideas]
We've long discussed spinning off integrations into their own repositories.  We have precedents in jdk-http-client, netty, finagle, and armeria, each of which began life separately.

We are facing new pressure to split off modules already here:
- Some integrations (e.g., play-json, twirl) have no plans to support Dotty.
- Servlet 5.0 is a breaking change, and its Jetty implementation requires Java 11
- Circe 0.14 is coming and we should be able to gracefully support it.

A problem for users is that most http4s modules currently have version, and now patch versions will vary. (imagine http4s-circe-0.22.3  on http4s-core-0.22.5).  This hasn't been a problem for http4s-jdk-http-client users.  Doing this in 0.21 would be traumatic, but starting in 0.22 and 1.0, we might be able to document our way out of it.

A hassle for us is that series/0.21, for as long as it's maintained, will stop merging cleanly to series/0.22 and beyond.

A huge benefit for us is that integrations can sink or swim on contributor interest.  Does anybody still care about http4s-json4s?  Help merge Scala Steward requests!  Does nobody care?  Let it stagnate!

I propose this points to splitting shortly before the v0.22.0-RC cycle, and a major curtailing of maintenance activity on series/0.21 at the same time.  And integrations that don't support Dotty when everything else does may go first.

**Comment by rossabaker** [comment]
~~Proposed repostiory layout~~ A better proposal is [here](https://github.com/http4s/http4s/discussions/4140#discussioncomment-261676).

* http4s
  * http4s-core
  * http4s-laws
  * http4s-testing (stop publishing)
  * http4s-tests (unpublished)
  * http4s-server
  * http4s-client
* http4s-dsl
  * http4s-dsl
* http4s-prometheus-metrics
  * http4s-prometheus-metrics 
* http4s-dropwizard-metrics
  * http4s-dropwizard-metrics 
* http4s-ember
  * http4s-ember-core
  * http4s-ember-server
  * http4s-ember-client
* http4s-blaze
  * http4s-blaze-core
  * http4s-blaze-server
  * http4s-blaze-client
* http4s-async-http-client
  * http4s-async-http-client
* http4s-servlet
  * http4s-servlet  
* http4s-jetty
  * http4s-jetty-client
  * http4s-jetty-server
* http4s-okhttp-client
  * http4s-okhttp-client
* http4s-tomcat
  * http4s-tomcat
* http4s-jawn
  * http4s-jawn
* http4s-argonaut
  * http4s-argonaut
* http4s-boopickle
  * http4s-boopickle
* http4s-circe
  * http4s-circe
* http4s-json4s
  * http4s-json4s
  * http4s-json4s-native
  * http4s-json4s-jackson
* http4s-play-json
  * http4s-play-json
* http4s-scala-xml
  * http4s-scala-xml
* http4-scalatags
  * http4s-scalatags
* http4s.org
  * http4s-website

http4s-bench could, like, scatter.

Projects could have their own mdoc sites.  "Integration" docs are tricky, if we want to show an end-to-end example of a server backend, some middlewares, and a JSON codec.  The docs project currently depends on the kitchen sink.  Examples projects have the same conundrum.

Scalafix is, um, split into the fixed modules?

**Reply by rossabaker** [reply]
Some of these I care about and some of these, I, well, don't.  Rather than pre-emptively establishing each repository, we should consider a call for a maintainer of each.

**Reply by rossabaker** [reply]
Some historical precedent on satellite repositories: argonaut used to break faster than we did.  We would support one version of argonaut in the main repo, and we had another http4s-argonaut-6-1 or whatever for the disfavored version.  The satellite versions were poorly maintained.

Reasons we might do better this time:
- argonaut had no fans among the http4s maintainers
- there was no scala-steward
- release automation was not as good as we have now

Maintainership is key for this to be a success, but I'm also less worried: if absolutely nobody cares enough to maintain a module, why should the core maintainers keep dragging it along?

**Comment by RaasAhsan** [comment]
1. How does this interact with the three-branch strategy? Is there going to be a `main` and `series/0.22.x` on every repo?
2. What is the expectation around compatibility between version numbers of repositories?

**Reply by rossabaker** [reply]
> How does this interact with the three-branch strategy? Is there going to be a main and series/0.22.x on every repo?

I think yes, each satellite repo would need a parallel branch structure.  (I think most projects should have this: production branch(es) and development.)

Multiple versions of an integration complicate things.  If we chose to support Servlet 3 and Servlet 5 concurrently, would it be 2x branches?  I don't want to do that.  But in this proposal, we could upgrade to Servlet 5 "someday" without a corresponding breaking change to http4s-core.

> What is the expectation around compatibility between version numbers of repositories?

I think to start with, 0.22 -> 0.22 and 1.0 -> 1.0.  But since breaking changes can be made in a module without breaking the entire ecosystem, this correspondence is not guaranteed to last.  (Thought exercise: what happens if circe-0.15 falls out of the sky the week after we release http4s-core-0.22.0?)

**Comment by rossabaker** [comment]
_Edit: September-November 2020, Scala 2.12 and 2.13_:

| Module             | Downloads | Incremental third-party dependencies                                       |
|--------------------|----------:|----------------------------------------------------------------------------|
| core               |      100% | case-insensitive, cats, cats-effect, cats-parse, log4s, scodec-bits, vault |
| blaze-core         |       86% | blaze                                                                      |
| client             |       86% |                                                                            |
| dsl                |       82% |                                                                            |
| jawn               |       76% | jawn, jawn-fs2                                                             |
| circe              |       76% | circe                                                                      |
| blaze-client       |       73% |                                                                            |
| blaze-server       |       71% |                                                                            |
| server             |       70% |                                                                            |
| async-http-client  |       17% | async-http-client                                                          |
| prometheus-metrics |       12% | prometheus-client                                                          |
| scala-xml          |        7% | scala-xml                                                                  |
| twirl              |        6% | twirl                                                                      |
| servlet            |        6% | servlet-api                                                                |
| jetty-server       |        6% | jetty                                                                      |
| okhttp-client      |        6% | okhttp                                                                     |
| testing            |        5% |                                                                            |
| argonaut           |        4% | argonaut                                                                   |
| tomcat             |        4% | tomcat                                                                     |
| json4s             |        4% | json4s                                                                     |
| json4s-jackson     |        4% |                                                                            |
| json4s-native      |        4% |                                                                            |
| laws               |        4% | discipline, scalacheck                                                     |
| dropwizard-metrics |        4% | dropwizard                                                                 |
| ember-core         |        3% | log4cats                                                                   |
| play-json          |        3% | play-json                                                                  |
| scalatags          |        3% | scalatags                                                                  |
| ember-client       |        3% | keypool                                                                    |
| jetty-client       |        3% |                                                                            |
| boopickle          |        3% | boopickle                                                                  |
| ember-server       |        3% |                                                                            |

Analysis to follow.

**Reply by rossabaker** [reply]
## Important

### http4s
* http4s-core
* http4s-laws 
* http4s-testing
* http4s-tests
* http4s-server
* http4s-client
* http4s-ember-core
* http4s-ember-client
* http4s-ember-server
* http4s-dsl
* http4s-jawn
* http4s-circe
* http4s-docs
* http4s-examples

Batteries included repository.  Is enough to provide an example and documentation of a traditional RESTful API.

We could try to replace log4s with cats in this repo and reduce liability.  log4s still fits better in blaze.

This repository would still have a lot of dependency liabilities.  A new circe would be an API-breaking change, but historically, they've changed on about our cadence.

### blaze
* http4s-blaze-core
* http4s-blaze-server
* http4s-blaze-client

It's audacious to replace the 75% backends with the 3% backend, but blaze has been privileged since 2013 and ember is new and unpromoted.  We need to make the world we want.  Also, this one has caused us a lot of pain in CI.

At the same time, it's synonymous with our success, and we need to keep it alive until something else is as trusted.

Will be moved into the blaze repo.  See comment in this thread.

### http4s-servlet
* http4s-servlet 

Usage is low, but this is how we've gotten FP into The Enterprise.  I don't expect anyone who hasn't lived it to understand, but this is important to me, and is appreciated by the few like me.  And it's something most of the rest of Scala can't offer.

## Redundant or niche

More than happy to support these, but Someone™ should take ownership.

### http4s-prometheus-metrics
* http4s-prometheus-metrics 

I've gotten this out of The Clouds for a long time, but a lot of people are using it?

### http4s-async-http-client
* http4s-async-http-client

Has a robust following due to frustration with blaze-client.

### http4s-okhttp-client
* http4s-okhttp-client

### http4s-jetty 
* http4s-jetty
* http4s-jetty-client
* http4s-jetty-server

There is the jetty4s project.  We should evalulate a reunification or adoption.

### http4s-tomcat
* http4s-tomcat

Jetty serves this role well.

### http4s-dropwizard-metrics
*  http4s-dropwizard-metrics 

Prometheus is winning?

### http4s-boopickle
* http4s-boopickle

I think @cquiroz uses this.

### http4s-play-json
* http4s-play-json

~~Lack of commitment to Dotty is concerning.~~ There is new activity! :tada:

### http4s-scala-xml
* http4s-scala-xml

I could be persuaded to keep this in core, but I'm old enough to remember when it mattered.  I won't let this die.

### http4-scalatags
* http4s-scalatags

Dotty PR is being ignored.  Concerning.

### http4s-twirl
* http4s-twirl

Lack of commitment to Dotty is concerning.

## Disrecommended

### http4s-argonaut
* http4s-argonaut

Circe forked it, improved it, and gave it a Code of Conduct.  There's no reason to use this.

### http4s-json4s
* http4s-json4s
* http4s-json4s-jackson
* http4s-json4s-native

Reflection based.  Just does not fit.

## Website

### http4s.org
* http4s-website

The home for generic project info.

**Reply by RaasAhsan** [reply]
Thoughts on pulling the `blaze` repo in with http4s-blaze? Not sure how many people use blaze on its own outside http4s but could reduce the burden of having to juggle through 2-3 repos

**Reply by rossabaker** [reply]
I considered that.  It's very much against the original author's wishes, and blaze hasn't seen a breaking change in a very long time because of its few dependencies.  But it's also confusing to have two, and Blaze never did take off outside http4s.  It's been 2.5 years since the original author's last commit, and nothing prevents Bryce or anyone from forking off the present state and running with the original vision.  Yeah, let's flatten it.

**Reply by isomarcte** [reply]
I'm generally in favor of this layout.

**Reply by rossabaker** [reply]
Am I the only one who doesn't hate XML?

**Reply by cquiroz** [reply]
I’m ok with xml for data, not a fun for builds

**Reply by rossabaker** [reply]
In retrospect, I'm not sure I still think that circe, jawn, or dsl belong here, though I think dsl would be a monster to remove due to tests.  

Nothing says the schism needs to complete in one fell swoop, and what was laid out here is about to be materialized before 0.23.12.

**Reply by armanbilge** [reply]
I bet @bplommer can scalafix the DSL away :P

**Reply by armanbilge** [reply]
> though I think dsl would be a monster to remove due to tests.

Btw, nobody says we have to _remove_ the DSL from the core repo after we schism it. We just have to stop publishing it (and possibly merge it into the testing support project). From there it could eventually rot out of existence.

**Comment by ChristopherDavenport** [comment]
Ember has a volunteer if that was not obvious.

**Reply by rossabaker** [reply]
Do you prefer it is a side repo or is built-in?  Blaze is where it is "by default," and Ember is a more modern default, but you've also been bummed by compatibility constraints in the past.  I'm gathering a few more numbers and about to post another idea that keeps Ember here.

**Reply by ChristopherDavenport** [reply]
I think we could either make them all independent, or keep blaze/ember together as a default pair.

**Reply by rossabaker** [reply]
Oh, do you think evicting blaze is too aggressive?  It puts a spotlight on Ember, I think for better in the long term and for worse in the short term.

**Reply by ChristopherDavenport** [reply]
I'm happy to evict blaze and leave ember in the spotlight. Does put a bit more pressure on, but I think we're approaching a strong place. The question is things like Date,etc where parsing is currently our bottleneck.

**Reply by armanbilge** [reply]
👍 to this. Now that ember is also the server/client for Node.js (and blaze will never cross).

**Comment by amesgen** [comment]
I don't know whether this has already been discussed, but what about publishing the abstract testing stuff which is currently `test` scoped (like [`ClientRouteTestBattery`](https://github.com/http4s/http4s/blob/f1a4ab056aee18cb5927ff45ee8bdd14bba5366e/client/src/test/scala/org/http4s/client/ClientRouteTestBattery.scala). This would be even more important after the Big Split, but even now, http4s-jdk-http-client [would benefit](https://github.com/http4s/http4s-jdk-http-client/blob/2978b35aee3403a46438f519db67dcbfcff8f5c3/core/src/test/scala/org/http4s/client/jdkhttpclient/JdkHttpClientSpec.scala) if it could use `ClientRouteTestBattery`.

**Reply by rossabaker** [reply]
They're not really "laws", but maybe we could rebrand them as such.  We already publish those, but they don't depend on client.  What we really want is more of a "test compatibility kit" -- an http4s backend behaves like ____.

I think in one form or another, it's a good idea and increasingly necessary.

**Reply by amesgen** [reply]
FTR I will try to put a PR together for this when CE3/Dotty are (more or less) settled.

**Comment by rossabaker** [comment]
I expect our first Dotty milestone to be ready within the next few days, and I don't expect Twirl or Scalatags to have a release.  We're going to need to figure out what to do about those real soon.

**Reply by SethTisue** [reply]
(I do still expect Twirl will make the jump, but the time frame is still "hopefully in the next couple months"-ish)

**Reply by rossabaker** [reply]
Thanks, @SethTisue!

For now it requires a gross hack, but we can keep it compiling and just not published for Dotty. As long as there is good intent and no obvious alternative, I'm inclined to try to keep modules alive.  I think both conditions are met here.

```scala
      libraryDependencies.value.map {
        case module if module.name == "twirl-api" =>
          module.withDottyCompat(scalaVersion.value)
        case module => module
      }
```

**Comment by armanbilge** [comment]
I'm curious how the new JS-only modules will fit into this. Namely:
* `node-serverless`: adds no new dependencies
* `dom-{core,fetch-client,service-worker}`: only dependency is `scala-js-dom`, a facade for browser APIs, which has historically been stable

Why they should stay in the main repo:
* small/lightweight
* tiny surface area for change
* canonical implementations for their respective platforms (e.g., fetch is _the_ native browser client)

Why they should leave:
* the majority of use cases for `node-serverless` may get absorbed into [typelevel/feral](https://github.com/typelevel/feral)
* all modules in the [current proposal](https://github.com/http4s/http4s/discussions/4140#discussioncomment-261676) cross-build for JVM/JS. So, these JS-only modules would stick out a bit

**Reply by rossabaker** [reply]
I don't have enough Scala.JS experience to add much to the answers, but I think you're asking the right questions.  Basically, we don't want anything that makes a breaking change on a different cycle than us, or is likely to lag publishing when there's a new compiler.

**Reply by armanbilge** [reply]
Yep, I think we're looking at `scala-js-dom` which fell behind with Scala 3 support. I'm now a maintainer for that repo although when is this new compiler rolling around? 😆

**Reply by rossabaker** [reply]
Well, I have no idea on Scala.JS.  I was thinking about breaking versions of scala-library, but this is not the time of day to allow myself to get riled up.

**Reply by armanbilge** [reply]
Hmm, good point about the Scala.js compiler. That's now built-in to dotty, and I doubt they'll make a breaking change to the compiler plugin for Scala 2.x. They just released Scala.js 1.7 with no known bugs.

**Reply by armanbilge** [reply]
Update on this: the dom modules are moved out, ultimately due to the build/test environment complexities they require.

That only leaves `node-serverless`. I think this one should stay, largely in part because it turned out to be useful for implementing the scaffold server used in the test suite.

**Reply by rossabaker** [reply]
I'm going to need to spend some time understanding the new dependencies.  Any time we use one of our backends to test the others, we have to be careful of circular dependencies (which the build should catch), or coupling that makes things harder to test later.  That's why some of the JVM servers are implemented on the raw Sun server.  (Previously on raw Jetty, but even that caused problems as we started thinking about Servlet 5!)

I think we're in at least a good interim place.

**Reply by armanbilge** [reply]
Yup, I'm mindful of this :) to avoid precisely this problem the node-serverless module no longer has tests (it is tested implicitly when ember-client itself tests against it). It uses Node's built-in server so is equivalent to the raw Sun server.

**Comment by rossabaker** [comment]
For http4s-servlet, I ran:

```console
$ git filter-repo --force --path servlet --path testing --path examples/src --path examples/war --replace-message ../expressions.txt
```

To rewrite PR references, I used this `../expressions.txt`.  It may be too broad, but it worked here:

```txt
#==>http4s/http4s#
```

With the result on remote branch `core/filtered`, I ran on the new repo:

```console
$ git merge --allow-unrelated-histories core/filtered
```

And then I copied relevant parts of the build.

**Reply by armanbilge** [reply]
I wish I had done this when we split http4s-dom, but it didn't have much history at the time anyway.

**Reply by rossabaker** [reply]
It doesn't produce a buildable history, but that's still in core, and it preserves proper credit.

**Reply by armanbilge** [reply]
Ok, I just gave this a try in https://github.com/http4s/http4s-scala-xml which has just published a snapshot :)

A couple notes from me:
1. `git filter-repo` needs https://github.com/newren/git-filter-repo
2. Since keeping tags seems kind of non-sensical, MiMa needs to be configured manually like so:
    ```scala
    ThisBuild / tlMimaPreviousVersions ++= (0 to 11).map(y => s"0.23.$y").toSet
    ```

Otherwise Ross's instructions are good and it was easy.

**Reply by rossabaker** [reply]
Right, I neglected to mention I'm deleting the tags that come into the new repo.  `git tag -d $(git tag -l)`.

**Comment by rossabaker** [comment]
One thing we missed was changes from 0.23 -> 1.0 in the spun off repos.  Those changes are lost and not currently in a cherry-pickable form.  Blaze is going to be the worst of these.  I'll try to figure out a pattern and post it here.

**Reply by armanbilge** [reply]
I think we can just run git-filter-repo again on `main`. If you look at http4s-scala-xml and http4s-scala-xml-1 they actually have identical commits e.g.:
https://github.com/http4s/http4s-scala-xml-1/commit/dd25fc4c759ef3e89c5deb410345722b84a07ce8
https://github.com/http4s/http4s-scala-xml/commit/dd25fc4c759ef3e89c5deb410345722b84a07ce8

**Reply by rossabaker** [reply]
They would have to be filtered with the same arguments.  Do you have what you did for blaze?  Mine didn't match.

**Reply by armanbilge** [reply]
Crap. I reckon it was something like:
```
git filter-repo --force --path blaze-client/ --path blaze-core/ --path blaze-server/ --path examples/src/ --path examples/blaze/ --replace-message ../expressions.txt
```

But I was also unable to replicate it 🤔

**Reply by rossabaker** [reply]
The `blaze/` directory might have been in there somewhere.  That's where it all originated.

If you don't have it in your shell history, I can figure something else out for this one.

**Reply by armanbilge** [reply]
I do all my OSS in ephemeral cloud machines, so that history is long gone 😕 I don't think I put `blaze/` though, I actually didn't know about that history, hope I didn't muck it up 😬

**Reply by rossabaker** [reply]
If I run your command from b30ad207c46142008482aadde2cc9b89158ff9dd, it's reproducible.  If I run it from main, it's not.  I can try to cherry-pick them, but I fear I'll be reliving a lot of merge conflicts.  Worst case, I can merge an alternate view of the history.  Oof.

**Reply by rossabaker** [reply]
Sorted in https://github.com/http4s/blaze/pull/683

**Comment by rossabaker** [comment]
Several of these ancillary repos are starting to get an 0.23 and a 1.0 branch.  I know we've done work to versioned docs, but I am not sure on the details.  We need to figure out whether to set that up, or just pick a documented branch.

**Reply by armanbilge** [reply]
The jdk-http-client is the current best example of how to set this up.
https://github.com/http4s/http4s-jdk-http-client/blob/e19038c93b15dbd38e86eed8ab144d9f9544532c/build.sbt#L91-L104

**Reply by armanbilge** [reply]
I have a new proposal for how to do versioned documentation, that is easier to set up. See:
- https://github.com/typelevel/sbt-typelevel/pull/317

**Comment by rossabaker** [comment]
I'm releasing http4s-scala-xml off main before releasing off series/0.23.  This defeats our "linearized releases" strategy to dedupe the changelog.  Maybe that's okay.
